pkg/image/admission/imagepolicy/rules/rules.go

```diff
@@ -0,0 +1,143 @@
+package rules
+
+import (
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/util/sets"
+
+	"github.com/openshift/origin/pkg/image/admission/imagepolicy/api"
+	imageapi "github.com/openshift/origin/pkg/image/api"
+)
+
+type ImagePolicyAttributes struct {
+	Resource      unversioned.GroupResource
+	Name          imageapi.DockerImageReference
+	OriginalName  string
+	Image         *imageapi.Image
+	ExcludedRules sets.String
+}
+
+type RegistryMatcher interface {
+	Matches(name string) bool
+}
+
+type RegistryNameMatcher imageapi.DefaultRegistryFunc
+
+func (m RegistryNameMatcher) Matches(name string) bool {
+	current, ok := imageapi.DefaultRegistryFunc(m)()
+	if !ok {
+		return false
+	}
+	return current == name
+}
+
+type nameSet []string
+
+func (m nameSet) Matches(name string) bool {
+	for _, s := range m {
+		if s == name {
+			return true
+		}
+	}
+	return false
+}
+
+func NewRegistryMatcher(names []string) RegistryMatcher {
+	return nameSet(names)
+}
+
+type resourceSet map[unversioned.GroupResource]struct{}
+
+func (s resourceSet) addAll(other resourceSet) {
+	for k := range other {
+		s[k] = struct{}{}
+	}
+}
+
+func prepareSourceRule(rule *api.ImageCondition) (requiresImage bool, covers resourceSet) {
+	switch {
+	case len(rule.MatchImageLabels) > 0,
+		len(rule.MatchImageAnnotations) > 0,
+		len(rule.MatchSignatures) > 0,
+		len(rule.MatchDockerImageLabels) > 0:
+		requiresImage = true
+	}
+
+	covers = make(resourceSet)
+	for _, gr := range rule.OnResources {
+		covers[gr] = struct{}{}
+	}
+
+	return requiresImage, covers
+}
+
+// emptyImage is used when resolution failures occur but resolution failure is allowed
+var emptyImage = &imageapi.Image{
+	DockerImageMetadata: imageapi.DockerImage{
+		Config: &imageapi.DockerConfig{},
+	},
+}
+
+func matchSourceRule(rule *api.ImageCondition, integrated RegistryMatcher, attrs *ImagePolicyAttributes) bool {
+	if rule.MatchIntegratedRegistry && !integrated.Matches(attrs.Name.Registry) {
+		return false
+	}
+	if len(rule.MatchRegistries) > 0 && !hasAnyMatch(attrs.Name.Registry, rule.MatchRegistries) {
+		return false
+	}
+
+	// all subsequent calls require the image
+	image := attrs.Image
+	if image == nil {
+		if !rule.AllowResolutionFailure {
+			return false
+		}
+		// matches will be against an empty image
+		image = emptyImage
```

Comment:
There is an important distinction between "try to resolve" and "fail if you
can't resolve".  A rule that tries to resolve, has ignore failure, and no
conditions, is a valid state.  Resolution can't be force on or force off,
we have to allow resolution to be best effort.

On Fri, Aug 5, 2016 at 9:14 AM, David Eads <notifications@github.com> wrote:

> In pkg/image/admission/imagepolicy/rules/rules.go
> <https://github.com/openshift/origin/pull/8995#discussion_r73689591>:
>
> > +func matchSourceRule(rule *api.ImageCondition, integrated RegistryMatcher, attrs *ImagePolicyAttributes) bool {
> > +	if rule.MatchIntegratedRegistry && !integrated.Matches(attrs.Name.Registry) {
> > +		return false
> > +	}
> > +	if len(rule.MatchRegistries) > 0 && !hasAnyMatch(attrs.Name.Registry, rule.MatchRegistries) {
> > +		return false
> > +	}
> > +
> > +	// all subsequent calls require the image
> > +	image := attrs.Image
> > +	if image == nil {
> > +		if !rule.AllowResolutionFailure {
> > +			return false
> > +		}
> > +		// matches will be against an empty image
> > +		image = emptyImage
>
> This doesn't make any sense to me. If you don't have the information to
> continue, you don't have it. Return here.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/8995/files/f96f5e56ba70c7c0f24f875621efdf1ef45ce3c0#r73689591>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p5QgWGEdioYHRVCF_Ot0HsG50AVMks5qczcggaJpZM4IlDlb>
> .
>